ext/standard/array.c

```diff
@@ -3694,13 +3694,11 @@ PHPAPI int php_array_merge_recursive(HashTable *dest, HashTable *src) /* {{{ */
 				ZVAL_DEREF(src_zval);
 				ZVAL_DEREF(dest_zval);
 				thash = Z_TYPE_P(dest_zval) == IS_ARRAY ? Z_ARRVAL_P(dest_zval) : NULL;
-				if ((thash && GC_IS_RECURSIVE(thash)) || (src_entry == dest_entry && Z_ISREF_P(dest_entry) && (Z_REFCOUNT_P(dest_entry) % 2))) {
+				if (thash && GC_IS_RECURSIVE(thash)) {
 					zend_throw_error(NULL, "Recursion detected");
 					return 0;
 				}
 
-				ZEND_ASSERT(!Z_ISREF_P(dest_entry) || Z_REFCOUNT_P(dest_entry) > 1);
-				SEPARATE_ZVAL(dest_entry);
```

Comment:
We can remove this one (move it after `convert_to_array`) because `convert_to_array` will change `dest_entry` itself, not its value.

This function needs some cleanup, as there are leftovers from previous refactorings. For instance I believe that `src_entry == dest_entry` above is always false, and we don't use the result of `ZVAL_DEREF(dest_zval)` because of `dest_zval = dest_entry` below. However this is out of scope of this bug fix.